### ISSUE TITLE
Fix AsymmetricSignatureProvider to support RSA keys on CoreCLR/Linux

### ISSUE THREAD

**Issue by PinpointTownes** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#324]
Moved from https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/179#issuecomment-155972405.

_____________

On CoreCLR, `AsymmetricSignatureProvider` tries to use `RSACng` (which is Windows-specific) instead of `RSAOpenSsl`.

      System.DllNotFoundException: Unable to load DLL 'ncrypt.dll': The specified module could not be found.
       (Exception from HRESULT: 0x8007007E)
         at Interop.NCrypt.NCryptOpenStorageProvider(SafeNCryptProviderHandle& phProvider, String pszProviderName, Int32 dwFlags)
         at Internal.Cryptography.Helpers.OpenStorageProvider(CngProvider provider)
         at System.Security.Cryptography.CngKey.Import(Byte[] keyBlob, CngKeyBlobFormat format, CngProvider provider)
         at System.Security.Cryptography.RSACng.ImportParameters(RSAParameters parameters)
         at System.IdentityModel.Tokens.AsymmetricSignatureProvider.ResolveDotNetCoreAsymmetricAlgorithm(AsymmetricSecurityKey key, String algorithm, Boolean willCreateSignatures)
         at System.IdentityModel.Tokens.AsymmetricSignatureProvider..ctor(AsymmetricSecurityKey key, String algorithm, Boolean willCreateSignatures)
         at System.IdentityModel.Tokens.SignatureProviderFactory.CreateProvider(SecurityKey key, String algorithm, Boolean willCreateSignatures)
         at System.IdentityModel.Tokens.SignatureProviderFactory.CreateForSigning(SecurityKey key, String algorithm)
         at System.IdentityModel.Tokens.RsaSecurityKey.GetSignatureProvider(String algorithm, Boolean verifyOnly)
         at System.IdentityModel.Tokens.SecurityKey.GetSignatureProviderForSigning(String algorithm)
         at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.CreateJwt(String issuer, String audience, IEnumerable`1 claims, Nullable`1 notBefore, Nullable`1 expires, Nullable`1 issuedAt, SigningCredentials signingCredentials)
         at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.CreateToken(String issuer, String audience, ClaimsIdentity subject, Nullable`1 notBefore, Nullable`1 expires, Nullable`1 issuedAt, SigningCredentials signingCredentials)

**Comment by tushargupta51** [comment]
@PinpointTownes did you start seeing this issue in rc1 only? have you tried using the beta8 bits on linux/coreclr?

**Comment by PinpointTownes** [comment]
Though I haven't tried with beta8, I'm pretty sure it affects all versions, since `AsymmetricSignatureProvider` used `RSACryptoServiceProvider` in the first releases and `RSACng` starting with [this commit] (https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/commit/37b116d83a57e81397d25ecce8109066c378dc41).

And since `RSACryptoServiceProvider` and `RSACng` don't work on CoreCLR/Linux, I'd be extremely surprised to learn that raw RSA keys used to work on beta8.

**Comment by tushargupta51** [comment]
@PinpointTownes That makes sense. Since the issue was reported post rc1, I was wondering if it was working in earlier versions somehow. Until RSA.Create() shows up(https://github.com/dotnet/corefx/issues/2953), I will use `RSACng` or `RSAOpenSSL` based on the underlying os.

**Comment by PinpointTownes** [comment]
Yup, that's probably the best approach. I've used roughly the same one for ASOS:

```csharp
internal static RSA GenerateKey(IRuntimeEnvironment environment) {
    if (string.Equals(environment.OperatingSystem, "Windows", StringComparison.OrdinalIgnoreCase)) {
#if DNXCORE50
        // On CoreCLR, use RSACng.
        return new RSACng(2048);
#else
        // On desktop CLR, use RSACryptoServiceProvider.
        return new RSACryptoServiceProvider(2048);
#endif
    }

    // When the runtime is identified as Mono, use RSACryptoServiceProvider independently of the operating system.
    if (string.Equals(environment.RuntimeType, "Mono", StringComparison.OrdinalIgnoreCase)) {
        return new RSACryptoServiceProvider(2048);
    }

#if DNXCORE50
    // On Linux and Darwin, use RSAOpenSsl when running on CoreCLR.
    if (string.Equals(environment.OperatingSystem, "Linux", StringComparison.OrdinalIgnoreCase) ||
        string.Equals(environment.OperatingSystem, "Darwin", StringComparison.OrdinalIgnoreCase)) {
        return new RSAOpenSsl(2048);
    }
#endif

    // If no appropriate implementation can be found, throw an exception.
    throw new PlatformNotSupportedException("No RSA implementation compatible with your configuration can be found.");
}

```

**Comment by sunsided** [comment]
According to git my experiments with `X509Certificate2` where I stumbled across this problem on Mono/Linux are from October 20th using rc1 packages. However, for the time being, neither certificates nor dynamically generated keys (by ASOS, if I got that right) work for me with middle aged rc1 and current rc2 packages - on Mono/Linux, that is, I stopped trying on Windows.

**Comment by PinpointTownes** [comment]
Cross-platform support? Pff, who cares.

**Comment by sunsided** [comment]
Wait, what?

**Comment by tushargupta51** [comment]
@sunsided This issue is fixed in #328
